src/operations/blas3/gemm_no_local_full_vec.hpp

```diff
@@ -0,0 +1,911 @@
+   * @brief binding the placeholder accessors to the SYCL command group
+   * handler
+   * @param h: SYCL command group handler. */
+  void bind(cl::sycl::handler &h) {
+    a_.bind(h);
+    b_.bind(h);
+    c_.bind(h);
+  }
+
+  void adjust_access_displacement() {
+    a_.adjust_access_displacement();
+    b_.adjust_access_displacement();
+    c_.adjust_access_displacement();
+  }
+
+ private:
+  /*!
+   * @brief Loads a full block of A and multiple partial blocks of B, computing
+   * the result for each iteration.
+   * @tparam packet_size : Vector size
+   * @tparam check_boundary : True if block is external, false if not
+   * @tparam check_k : Whether to check in K dimension.
+   * @tparam BoundaryCheckM : Type of function for checking boundary in M
+   * dimension.
+   * @tparam BoundaryCheckN : Type of function for checking boundary in N
+   * dimension.
+   * @tparam PointerType : Type of the input pointers for A and B matrices.
+   * @param A : Input matrix A
+   * @param B : Input matrix B
+   * @param boundary_check_m : Function which checks boundary in M dimension.
+   * @param boundary_check_n : Function which checks boundary in N dimension.
+   * @param A_ptr_index : index of next element for A
+   * @param B_ptr_index : index of next element for B
+   * @param lda : leading dimension of A
+   * @param ldb : leading dimension of B
+   * @param k : the current value of K from the main loop which calls this
+   * method.
+   * @param reg_a : Pointer to private register for A.
+   * @param reg_b : Pointer to private register for B.
+   * @param reg_res : Pointer to private register for result.
+   * @param out_of_range : Controls whether to exit some functions early if
+   * block is out of range of A or B.*/
+  template <index_t packet_size, bool check_boundary, bool check_k,
+            typename BoundaryCheckM, typename BoundaryCheckN,
+            typename PointerType>
+  SYCL_BLAS_INLINE void load_and_compute_block(
+      PointerType A, PointerType B, BoundaryCheckM boundary_check_m,
+      BoundaryCheckN boundary_check_n, const index_t &A_ptr_index,
+      const index_t &B_ptr_index, const index_t &lda, const index_t &ldb,
+      const index_t &k, element_t *reg_a, element_t *reg_b, element_t *reg_res,
+      bool out_of_range
+#ifdef ARM_GPU
+      ,
+      const cl::sycl::nd_item<1> &id
+#endif
+  ) {
+    /*
+     * Loading a corresponding block of matrix A into reg_a
+     */
+    load_block_a<item_rows, packet_size, wg_rows * packet_size, check_boundary,
+                 check_k, packet_size, trans_a>(
+        A, reg_a, A_ptr_index, lda, boundary_check_m,
+        [=](const index_t &idx) SYCL_BLAS_ALWAYS_INLINE { return idx < k; },
+        out_of_range);
+#ifdef ARM_GPU
+    id.barrier(cl::sycl::access::fence_space::local_space);
+#endif
+
+#pragma unroll
+    for (int j = 0; j < packet_size; j++) {
+#pragma unroll
+      for (int i = 0; i < item_cols / packet_size; i++) {
+        const index_t ofs = i * B_ptr_index;
```

Comment:
It's possibly not worth spending lots of time on this but some SYCL compilers do not do loop strength reduction, so manually converting these index calculations to repeated additions may give better performance. This transform could probably be applied elsewhere as well.

```
index_t ofs = 0;
index_t col_ofs = 0;
for (int i = 0; i < item_cols / packet_size; i++) {
  ...
  ofs += B_ptr_index;
  col_ofs += (trans_b ? packet_size : 1) * wg_cols;
}
```

Reply:
I've applied this in the instance you specified there but have left the others alone for now. I'll make a note to track this for the future though